crates/macros/impl/src/cheatcodes.rs

```diff
@@ -0,0 +1,252 @@
+            };
+        }
+    })
+}
+
+/// Generates the `CHEATCODES` constant and implements `CheatcodeImpl` dispatch for an enum.
+fn derive_enum(name: &Ident, input: &syn::DataEnum) -> Result<TokenStream> {
+    if input.variants.iter().any(|v| v.fields.len() != 1) {
+        return Err(syn::Error::new(name.span(), "expected all variants to have a single field"))
+    }
+
+    // keep original order for matching
+    let variants_names = input.variants.iter().map(|v| &v.ident);
+
+    let mut variants = input.variants.iter().collect::<Vec<_>>();
+    variants.sort_by(|a, b| a.ident.cmp(&b.ident));
+    let variant_tys = variants.iter().map(|v| {
+        assert_eq!(v.fields.len(), 1);
+        &v.fields.iter().next().unwrap().ty
+    });
+    Ok(quote! {
+        /// All the cheatcodes in [this contract](self).
+        pub const CHEATCODES: &'static [&'static Cheatcode<'static>] = &[#(<#variant_tys as CheatcodeDef>::CHEATCODE,)*];
+
+        #[cfg(feature = "impls")]
+        impl #name {
+            pub(crate) fn apply<DB: crate::impls::DatabaseExt>(&self, ccx: &mut crate::impls::CheatsCtxt<'_, DB>) -> crate::impls::Result {
+                match self {
+                    #(Self::#variants_names(c) => crate::impls::Cheatcode::apply_traced(c, ccx),)*
+                }
+            }
+        }
+    })
+}
+
+fn check_named_fields(data: &DataStruct, ident: &Ident) {
+    for field in data.fields.iter() {
+        if field.ident.is_none() {
+            emit_warning!(ident, "all params must be named");
+        }
+    }
+}
+
+/// Flattens all the `#[doc = "..."]` attributes into a single string.
+fn get_docstring(attrs: &[syn::Attribute]) -> String {
+    let mut doc = String::new();
+    for attr in attrs {
+        if !attr.path().is_ident("doc") {
+            continue
+        }
+        let syn::Meta::NameValue(syn::MetaNameValue {
+            value: syn::Expr::Lit(syn::ExprLit { lit: syn::Lit::Str(s), .. }),
+            ..
+        }) = &attr.meta
+        else {
+            continue
+        };
+
+        let value = s.value();
+        if !value.is_empty() {
+            if !doc.is_empty() {
+                doc.push('\n');
+            }
+            doc.push_str(&value);
+        }
+    }
+    doc
+}
+
+/// Returns `(signature, hex_selector, declaration, description)` from a given `sol!`-generated
+/// docstring for a function.
+///
+/// # Examples
+///
+/// The following docstring (string literals are joined with newlines):
+/// ```text
+/// "Function with signature `foo(uint256)` and selector `0x1234abcd`."
+/// "```solidity"
+/// "function foo(uint256 x) external view returns (bool y);"
+/// "```"
+/// "Description of the function."
+/// ```
+///
+/// Will return:
+/// ```text
+/// (
+///     "foo(uint256)",
+///     "0x1234abcd",
+///     "function foo(uint256 x) external view returns (bool y);",
+///     "Description of the function."
+/// )
+/// ```
+fn func_docstring(doc: &str) -> (&str, &str, &str, &str) {
```

Comment:
should this be upstreamed

Reply:
This is just a hack to get information between `sol!` internals and cheatcode macro internals, I figured it's good enough